src/main/java/org/simplify4u/plugins/sign/openpgp/PGPSecretKeyUtils.java

```diff
@@ -133,4 +135,30 @@ private static String fingerprintToString(byte[] bytes) {
         }
         return ret.toString();
     }
+
+    /**
+     * Verify expiration time of secret key.
+     *
+     * @param secretKey     a key to check
+     * @param secretKeyRing a keyRing used for prepare message
+     *
+     * @throws PGPSignerException if key expired
+     */
+    public static void verifyKeyExpiration(PGPSecretKey secretKey, PGPSecretKeyRing secretKeyRing) {
+
+        long validSeconds = secretKey.getPublicKey().getValidSeconds();
```

Comment:
I wonder if that is always correct. There is a method `PGPSecretKey.replacePublicKey` that would set another public key with a different valid seconds number into the same secret key. Effectively our code here would see the new expiration date. Is that correct?

Reply:
I don't know what for is method `PGPSecretKey.replacePublicKey` but in our code it is not call. We read secret key from file and don't change content of it. So it will be always correct.

Reply:
My fear actually is not our code calling that on its own, but whether that other external tools modify the key file in a way that `replacePublicKey` does, and I wonder what the outcome is: Will our code still see the expiration date of the *secret* key, or will it now see the expiration date of that *new public* key? As this question has impact on the correctness of our validation check, maybe we should ask at the Bouncy Castle user list to be on the safe side?

Reply:
We can try to verify internal key signature, each `uid` and `subkey` has self signed signature, but we are in the same situation if somebody can change file with key probably generate new valid self signature.

Another idea is verify system access level to key file - if only is writable / readable by owner.

I will assume that key file is protected correctly by operating system.

Reply:
Before hastily hacking in something that we possibly don't need at all, I have asked our questions at [Bouncy Castle](https://github.com/bcgit/bc-java/issues/861). If you like, we can merge the PR as is and I will file the open questions as an open issue in our tracker?

Reply:
ok, we can open issue on our list and link with bc